modules/platforms/cpp/ignite/client/detail/cancellation_token_impl.cpp

```diff
@@ -0,0 +1,121 @@
+        if (m_result) {
+            callback(ignite_result<void>{*m_result});
+        } else {
+            m_callbacks.push_back(std::move(callback));
+        }
+        return;
+    }
+
+    m_cancelled = true;
+    m_callbacks.push_back(std::move(callback));
+
+    if (m_actions.empty()) {
+        m_result = ignite_result<void>{};
+        for (auto &cb : m_callbacks) {
+            cb(ignite_result<void>{*m_result});
+        }
+        return;
+    }
+
+    auto results = std::make_shared<std::vector<ignite_result<void>>>();
+    auto results_mutex = std::make_shared<std::mutex>();
+    auto expected_results = m_actions.size();
+    for (auto &action : m_actions) {
+        action([this, results, results_mutex, expected_results] (ignite_result<void> res) {
+            bool all_done{false};
+            { // Mutex locking scope
+                std::lock_guard guard0(*results_mutex);
+                results->push_back(std::move(res));
+                all_done = results->size() == expected_results;
+
+                if (all_done) {
+                    bool error_found{false};
+                    std::stringstream msg_builder;
+                    for (auto &result : *results) {
+                        if (!result.has_error()) {
+                            continue;
+                        }
+
+                        if (!error_found) {
+                            msg_builder << "One or more cancel actions failed: " << result.error().what_str();
+                            error_found = true;
+                            continue;
+                        }
+
+                        msg_builder << ", " << result.error().what_str();
+                    }
+
+                    { // Result mutex locking scope.
+                        std::lock_guard guard1(m_mutex);
+                        if (error_found) {
+                            m_result = {ignite_error(msg_builder.str())};
+                        } else {
+                            m_result = ignite_result<void>{};
+                        }
+                    }
+                }
+            }
+
+            if (all_done) {
+                std::lock_guard guard0(m_mutex);
+                for (auto &cb : m_callbacks) {
+                    cb(ignite_result<void>{*m_result});
+                }
+            }
+        });
+    }
+}
+
+void cancellation_token_impl::add_action(std::shared_ptr<ignite_logger> logger,
+    std::function<void(ignite_callback<void>)> action) {
+    auto callback = [logger](ignite_result<void> res) {
+        if (res.has_error()) {
+            logger->log_warning(
+                "Cancellation action that was added after the token was already canceled: " + res.error().what_str());
+        }
+    };
+
+    if (is_cancelled()) {
+        action(callback);
+        return;
+    }
+
+    std::unique_lock lock(m_mutex);
+    if (is_cancelled()) {
+        lock.unlock();
+        action(callback);
+        return;
+    }
+
```

Comment:
If we remove the first check outside the lock then `m_cancelled` does not have to be an atomic. If you want to leave it in, it has to be an atomic.

Reply:
Yeah, it should probably be atomic anyway, as `is_cancelled()` is a public method.